### PR TITLE
refactor(pypi): move config setting processing to the macro

### PR DESCRIPTION
**Pull request by aignas**
Before hand we would pass around `whl_alias` struct instances and it
would go through rounds of starlark functions which would end up
rendered as BUILD.bazel files in the hub repository. This means that
every time the output of the said functions would change, the
`MODULE.bazel.lock` would also possibly change.

Since we now have much better unit testing framework, we start relying
on these functions being evaluated at loading phase instead of repo rule
phase, which means that we can make the `BUILD.bazel` files not change a
lot when the underlying code changes, this should make the code more
maintainable in the long run and unblocks the code to accept extra
things, like env marker values or similar.

Summary:
- Remove the `repo` field from `whl_alias`.
- Rename `whl_alias` to `whl_config_setting`
- Move the tests around
- Simplify the `pkg_aliases` tests to use `contains` instead of exact
  equality to make things less brittle.
- Simplify the `pkg_aliases` tests to use different mocks to make
  expectations easier to understand.
- Make `whl_config_setting` hashable by using tuples instead of lists.
- Adjust how we store the `whl_config_setting` in the PyPI extension
  and optimize the passing to the `hub_repository`.

This is needed for #2319 and #2423.

To be in future PRs:
* Remove the need to pass `osx_versions`, etc to the `pkg_aliases` macro.

### CODE REVIEW

File: python/private/pypi/pkg_aliases.bzl [hunk 1]
```diff
@@ -132,3 +145,253 @@ def pkg_aliases(
...
+         * A list of config settings that are generated by ./pip_config_settings.bzl
+         * The list of default version settings.
+    """
+    prefixes = []
+    suffixes = []
+    setting_supported_versions = {}
+
+    if filename.endswith(".whl"):
+        if (0, 0) in glibc_versions:
+            fail("Invalid version in 'glibc_versions': cannot specify (0, 0) as a value")
```
Comment (rickeylev) [inline python/private/pypi/pkg_aliases.bzl:288]: Same comment here about wording as above.

File: python/private/pypi/pkg_aliases.bzl [hunk 2]
```diff
@@ -132,3 +145,253 @@ def pkg_aliases(
...
+        for p in prefixes
+        for suffix, versions in setting_supported_versions.items()
+    }
+
+    if suffixes or target_platforms or versioned:
+        return [
+            ":is_cp{}{}_{}".format(python_version, p, s)
+            for p in prefixes
+            for s in suffixes or [_non_versioned_platform(p) for p in target_platforms or []]
+        ], versioned
```
Comment (rickeylev) [inline python/private/pypi/pkg_aliases.bzl:339]: Can you unroll this a bit? The nested list comps and `or`'d expression make this hard to read and I'm wary of remembering precedence correctly.

File: python/private/pypi/render_pkg_aliases.bzl [hunk 3]
```diff
@@ -62,33 +62,65 @@ which has a "null" version value and will not match version constraints.
...
+        if alias.filename:
+            config_setting = """whl_config_setting(\n{}\n)""".format(
+                render.indent("\n".join([
+                    "{} = {},".format(k, repr(v))
+                    for k, v in {
+                        "filename": alias.filename,
+                        "target_platforms": alias.target_platforms,
+                        "version": alias.version,
+                    }.items()
+                    if v
```
Comment (rickeylev) [inline python/private/pypi/render_pkg_aliases.bzl:79]: How about creating text_util.render_kwargs() ? I've had to write similar code to this before. Looks like this same logic is duplicated down below, too.

**Review by rickeylev — APPROVED** [review]

File: python/private/pypi/pkg_aliases.bzl (hunk 2 above)
Comment (aignas) [reply python/private/pypi/pkg_aliases.bzl:339]: OK, I will clean this up.

**Comment by aignas** [comment]
I need to revisit the error message wording here. Will do later.

**Pull request merged**